tests/kernel/update_kernel.pm

```diff
@@ -0,0 +1,209 @@
+    }
+
+    script_run("lsinitrd /boot/initrd-$kver-default | grep patch");
+    save_screenshot;
+    script_run("lsinitrd /boot/initrd-$kver-default | awk '/-patch-.*ko\$/ {print \$NF}' > /dev/$serialdev", 0);
+    ($module) = wait_serial(qr/lib*/) =~ /(^.*ko)\s+/;
+
+    mod_rpm_info($module);
+
+    script_run("uname -a");
+    save_screenshot;
+}
+
+sub install_lock_kernel {
+    my $version  = shift;
+    my @packages = qw(kernel-default  kernel-default-devel kernel-macros kernel-source);
+
+    # SLE12 and SLE12SP1 has xen kernel
+    if (!sle_version_at_least('12-SP2')) {
+        push @packages, qw(kernel-xen kernel-xen-devel);
+    }
+
+    # remove all kernel related packages from system
+    script_run("zypper -n rm " . join(' ', @packages), 700);
+
+    my @lpackages = @packages;
+    foreach (@packages) { s/$/-$version/ }
+
+    # install and lock needed kernel
+    zypper_call("in " . join(' ', @packages), exitcode => [0, 102, 103, 104]);
+    zypper_call("al " . join(' ', @lpackages));
+}
+
+sub prepare_kgraft {
+    my ($repo, $patch) = @_;
+    my $arch    = get_required_var('ARCH');
+    my $version = get_required_var('VERSION');
+    if (!sle_version_at_least('12-SP3')) {
+        $version = '12';
+    }
+
+    #install kgraft product
+    zypper_call("ar http://download.suse.de/ibs/SUSE/Products/SLE-Live-Patching/$version/$arch/product/ kgraft-pool");
+    zypper_call("ar http://download.suse.de/ibs/SUSE/Updates/SLE-Live-Patching/$version/$arch/update/ kgraft-update");
+    zypper_call("ref");
+    zypper_call("in sle-live-patching-release", exitcode => [0, 102, 103]);
+
+    #add repository with tested patch
+    zypper_call("ar $repo kgraft-test-repo");
+    my $kversion = script_output(q(zypper -n se -s kernel-default));
+    my $pversion = script_output(q(zypper -n se -r kgraft-test-repo -s));
+
+    #disable kgraf-test-repo for while
+    zypper_call("mr -d kgraft-test-repo");
+
+    my $wanted_version = right_kversion($kversion, $pversion);
+    fully_patch_system;
+    install_lock_kernel($wanted_version);
+    type_string("reboot\n");
+}
+
+sub right_kversion {
+    my ($kversion, $pversion) = @_;
+
+    my ($kver_fragment) = $pversion =~ qr/kgraft-patch-(\d+_\d+_\d+-\d+_*\d*_*\d*)-default/;
+    $kver_fragment =~ s/_/\\\./g;
+    my ($real_version) = $kversion =~ qr/($kver_fragment\.*\d*)/;
+
+    return $real_version;
+}
+
+sub update_kgraft {
+    my $patch = shift;
+    zypper_call("mr -e kgraft-test-repo");
+    zypper_call("ref");
+
+    if (!is_patch_needed($patch)) {
+        script_run(qq{rpm -qa --qf "%{NAME}-%{VERSION}-%{RELEASE} (%{INSTALLTIME:date})\n" | sort -t '-' > /tmp/rpmlist.before});
+        upload_logs('/tmp/rpmlist.before');
+
+        # Download HEAVY LOAD script
+        assert_script_run("curl -f " . autoinst_url . "/data/qam/heavy_load.sh -o /tmp/heavy_load.sh");
+
+        #run HEAVY Load script
+        script_run("bash /tmp/heavy_load.sh");
+        # warm up system
+        sleep 15;
```

Comment:
this part and the lines in before look like a duplication of tests/qam-kgraft/update_kgraft.pm. I recommend to extract common functionality.